src/cargo/ops/cargo_install.rs

```diff
@@ -316,40 +329,77 @@ fn install_one(
-            .into_iter()
-            .filter(|name| dst.join(name).exists())
-            .map(|name| (name, None))
-            .collect();
-        if !force && !duplicates.is_empty() {
-            let mut msg: Vec<String> = duplicates
-                .iter()
-                .map(|(name, _)| format!("binary `{}` already exists in destination", name))
-                .collect();
-            msg.push("Add --force to overwrite".to_string());
-            bail!("{}", msg.join("\n"));
-        }
-        Ok(duplicates)
-    };
-
     // WARNING: no_track does not perform locking, so there is no protection
     // of concurrent installs.
     if no_track {
         // Check for conflicts.
-        no_track_duplicates()?;
+        no_track_duplicates(&pkg, opts, &dst, force)?;
     } else if is_installed(&pkg, config, opts, &rustc, &target, root, &dst, force)? {
         let msg = format!(
             "package `{}` is already installed, use --force to override",
             pkg
         );
         config.shell().status("Ignored", &msg)?;
-        return Ok(false);
+        return Ok(None);
     }
 
+    Ok(Some(pkg))
+}
+
+fn no_track_duplicates(
+    pkg: &Package,
+    opts: &ops::CompileOptions,
+    dst: &Path,
+    force: bool,
+) -> CargoResult<BTreeMap<String, Option<PackageId>>> {
+    // Helper for --no-track flag to make sure it doesn't overwrite anything.
+    let duplicates: BTreeMap<String, Option<PackageId>> = exe_names(&pkg, &opts.filter)
+        .into_iter()
+        .filter(|name| dst.join(name).exists())
+        .map(|name| (name, None))
+        .collect();
+    if !force && !duplicates.is_empty() {
+        let mut msg: Vec<String> = duplicates
+            .iter()
+            .map(|(name, _)| format!("binary `{}` already exists in destination", name))
+            .collect();
+        msg.push("Add --force to overwrite".to_string());
+        bail!("{}", msg.join("\n"));
+    }
+    Ok(duplicates)
+}
+
+fn install_one<'cfg>(
+    config: &Config,
+    root: &Filesystem,
+    source_id: SourceId,
+    vers: Option<&str>,
+    opts: &ops::CompileOptions,
+    force: bool,
+    no_track: bool,
+    pkg: Package,
+) -> CargoResult<()> {
     config.shell().status("Installing", &pkg)?;
 
+    let dst = root.join("bin").into_path_unlocked();
+    let (mut ws, rustc, target) = make_ws_rustc_target(config, opts, &source_id, pkg.clone())?;
+
+    let mut td_opt = None;
+    let mut needs_cleanup = false;
+    if !source_id.is_path() {
+        let target_dir = if let Some(dir) = config.target_dir()? {
+            dir
+        } else if let Ok(td) = TempFileBuilder::new().prefix("cargo-install").tempdir() {
+            let p = td.path().to_owned();
+            td_opt = Some(td);
+            Filesystem::new(p)
+        } else {
+            needs_cleanup = true;
+            Filesystem::new(config.cwd().join("target-install"))
+        };
+        ws.set_target_dir(target_dir);
```

Comment:
Moved this tempdir logic into the compilation/installation function since it's so closely tied with the actual compilation process. We also needed `td_opt` after compilation. This made sense to me.